simpleline/render/renderer.py

```diff
@@ -0,0 +1,382 @@
+        # increment the error counter
+        if not self.input(active_screen.args, c):
+            self._input_error_counter += 1
+
+            # redraw the screen after 5 bad inputs
+            if self._input_error_counter >= 5:
+                self.redraw()
+            else:
+                self.input_required()
+
+    def raw_input(self, prompt, hidden=False):
+        """This method reads one input from user. Its basic form has only one
+        line, but we might need to override it for more complex apps or testing.
+        """
+        if self._input_thread is not None and self._input_thread.is_alive():
+            raise KeyError("Can't run multiple input threads at the same time!")
+
+        input_queue = queue.Queue()
+        self._input_thread = threading.Thread(target=self._thread_input, name="InputThread",
+                                              args=(input_queue, prompt, hidden))
+        self._input_thread.daemon = True
+        self._input_thread.start()
+        self._event_loop.process_signals(return_after=InputReadySignal)
+        return input_queue.get()  # return the user input
+
+    def input(self, args, key):
+        """Method called internally to process unhandled input key presses.
+
+        Also handles the main quit and close commands.
+
+        :param args: optional argument passed from switch_screen calls
+        :type args: anything
+
+        :param key: the string entered by user
+        :type key: str
+
+        :return: True if key was processed, False if it was not recognized
+        :rtype: True|False
+        """
+        # delegate the handling to active screen first
+        if not self._screen_stack.is_empty():
+            try:
+                key = self._screen_stack.pop(False).ui_screen.input(args, key)
+                if key == INPUT_PROCESSED:
+                    return True
+                elif key == INPUT_DISCARDED:
+                    return False
+            except ExitMainLoop:
+                raise
+            except Exception:    # pylint: disable=broad-except
+                self._event_loop.enqueue_signal(ExceptionSignal(self))
+                return False
+
+        # global refresh command
+        if not self._screen_stack.is_empty() and (key == Prompt.REFRESH):
+            self.redraw()
+            return True
+
+        # global close command
+        if not self._screen_stack.is_empty() and (key == Prompt.CONTINUE):
+            self.close_screen()
+            return True
+
+        # global quit command
+        elif not self._screen_stack.is_empty() and (key == Prompt.QUIT):
+            if self.quit_screen:
+                quit_screen = self.quit_screen
+                d = quit_screen(self, self._quit_message)
+                self.switch_screen_modal(d)
+                if d.answer:
+                    raise ExitMainLoop()
+            else:
+                raise ExitMainLoop()
+            return True
+
+        return False
```

Comment:
It looks like we can return `False` if the screen stack is empty, so we don't have to check it in every `if` condition.